### PR TITLE
Integrate Lite launcher into Lightning CLI

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

This PR changes the Lite CLI from argparse to click to be integrated into the main lightning CLI. Originally we though we can wrap the argparse CLI into click, but this doesn't work well. The simple solution seems to be the best. The click package has no dependencies.

```
lightning run --help
```
```
Usage: lightning run [OPTIONS] COMMAND [ARGS]...

  Run a Lightning application locally or on the cloud.

Options:
  --help  Show this message and exit.

Commands:
  app   Run an app from a file.
  lite  Run a Lightning Lite script.

```

The `lite` subcommand is only available if `lightning-lite` is installed. The name here is not final, it remains to be decided.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

cc @borda @carmocca @justusschock @awaelchli

### CODE REVIEW

File: src/lightning_lite/cli.py [hunk 1]
```diff
@@ -106,6 +119,7 @@ def _set_env_variables(args: Namespace) -> None:
     """
     os.environ["LT_CLI_USED"] = "1"
     os.environ["LT_ACCELERATOR"] = str(args.accelerator)
+    # TODO: what should be the default strategy if devices > 1?
```
Comment (justusschock) [inline src/lightning_lite/cli.py:122]: I'd go with ddp, which works in most cases.

**Review by justusschock — APPROVED** [review]

File: requirements/app/base.txt [hunk 2]
```diff
@@ -12,3 +12,4 @@ lightning-utilities==0.3.*
 beautifulsoup4<=4.8.2
 inquirer>=2.10.0
 psutil<5.9.4
+click>=8.0.0, <=8.1.3
```
Comment (carmocca) [inline requirements/app/base.txt:15]: How come the app reqs didn't set this before? :thinking:

File: requirements/lite/base.txt [hunk 3]
```diff
@@ -7,3 +7,4 @@ fsspec[http]>2021.06.0, <2022.6.0
 packaging>=17.0, <=21.3
 typing-extensions>=4.0.0, <=4.4.0
 lightning-utilities==0.3.*
+click>=8.0.0, <=8.1.3
```
Comment (carmocca) [inline requirements/lite/base.txt:10]: I don't think this should be added here. Lite should still work without click installed. The `CLI` interface is not a requirement for it's use.

File: src/lightning_lite/cli.py [hunk 4]
```diff
@@ -161,11 +174,10 @@ def _torchrun_launch(args: Namespace, script_args: List[str]) -> None:
...
-    args, script_args = _parse_args()
+def main(args: Namespace, script_args: Optional[List[str]] = None) -> None:
     _set_env_variables(args)
-    _torchrun_launch(args, script_args)
+    _torchrun_launch(args, script_args or [])
 
 
 if __name__ == "__main__":
-    main()
+    _run_model()
```
Comment (carmocca) [inline src/lightning_lite/cli.py:183]: Since click is only used when this is called, this should check that click is available etc

**Review by tchaton — APPROVED** [review]
Looks great ! Should we support `lightning run model --cloud`

File: requirements/app/base.txt (hunk 2 above)
Comment (awaelchli) [reply requirements/app/base.txt:15]: Some other package must be using it so it gets installed alongside of it.

File: requirements/lite/base.txt (hunk 3 above)
Comment (awaelchli) [reply requirements/lite/base.txt:10]: If click is not a requirement, what should I do? Ask user to install it if they use the CLI? At some point, we will have docs for Lite (and the standalone package) and we will show the command line as the primary way of interaction.

File: src/lightning_lite/cli.py [hunk 5]
```diff
@@ -13,38 +13,49 @@
...
 _log = logging.getLogger(__name__)
 
+_CLICK_AVAILABLE = RequirementCache("click")
+
 _SUPPORTED_ACCELERATORS = ("cpu", "gpu", "cuda", "mps", "tpu")
-_SUPPORTED_STRATEGIES = (None, "ddp", "dp", "deepspeed")
+_SUPPORTED_STRATEGIES = ("ddp", "dp", "deepspeed")
 _SUPPORTED_PRECISION = ("64", "32", "16", "bf16")
 
+if _CLICK_AVAILABLE:
```
Comment (carmocca) [inline src/lightning_lite/cli.py:33]: This will incur an import overhead. What if we put this entire definition inside `if name == main`?

**Review by carmocca — APPROVED** [review]

File: src/lightning_lite/cli.py [hunk 6]
```diff
@@ -161,11 +178,17 @@ def _torchrun_launch(args: Namespace, script_args: List[str]) -> None:
...
 
 
 if __name__ == "__main__":
-    main()
+    if not _CLICK_AVAILABLE:  # pragma: no cover
+        _log.error(
+            "To use the Lightning Lite CLI, you must have `click` installed."
+            " Install it by running `pip install -U click`."
+        )
+        exit(1)
```
Comment (carmocca) [inline src/lightning_lite/cli.py:192]:
```suggestion
        raise SystemExit(1)
```

**Review by carmocca — APPROVED** [review]

**Pull request merged**